### PR TITLE
[model] add model example GeniePath

### PR DESCRIPTION
**Pull request by kayzliu**
## Description

add model example GeniePath from paper [Geniepath: Graph neural networks with adaptive receptive paths](https://arxiv.org/abs/1802.00910). @zhjwy9343 

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [x] Code is well-documented
- [x] To the my best knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [x] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

File: examples/pytorch/geniepath/ppi.py [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+    # Step 4: training epochs =============================================================== #
+    for epoch in range(args.max_epoch):
+        model.train()
+        tr_loss = 0
+        tr_f1 = 0
+        num_blocks = 0
+        for subgraph in train_dataloader:
+            subgraph = subgraph.to(device)
+            label = subgraph.ndata['label'].to(device)
+            feat = subgraph.ndata['feat']
```
Comment (zhjwy9343) [inline examples/pytorch/geniepath/ppi.py:85]: Do we need to send the feat to device?

File: examples/pytorch/geniepath/ppi.py [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+            logits = model(subgraph, feat)
+
+            # compute loss
+            batch_loss = loss_fn(logits, label)
+            tr_loss += batch_loss.item()
+            tr_predict = np.where(logits.data.cpu().numpy() >= 0., 1, 0)
```
Comment (zhjwy9343) [inline examples/pytorch/geniepath/ppi.py:91]: ditto

File: examples/pytorch/geniepath/ppi.py [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+    num_blocks = 0
+    for subgraph in dataloader:
+        subgraph = subgraph.to(device)
+        label = subgraph.ndata['label'].to(device)
+        feat = subgraph.ndata['feat']
+        logits = model(subgraph, feat)
+
+        # compute loss
+        loss += loss_fn(logits, label).item()
+        predict = np.where(logits.data.cpu().numpy() >= 0., 1, 0)
```
Comment (zhjwy9343) [inline examples/pytorch/geniepath/ppi.py:25]: Normally for binary classification, we use torch.argmax to get the predict 0 or 1. Any specific reasons here are you using np.where?

File: examples/pytorch/geniepath/ppi.py (hunk 1 above)
Comment (kayzliu) [reply examples/pytorch/geniepath/ppi.py:85]: It's not necessary, cuz the subgraph is already on the device. Same as the line 84. to(device) deleted in line 84.

File: examples/pytorch/geniepath/ppi.py (hunk 3 above)
Comment (kayzliu) [reply examples/pytorch/geniepath/ppi.py:25]: PPI dataset has more than two classes, so I think np.where is necessary.

File: examples/pytorch/geniepath/ppi.py (hunk 2 above)
Comment (kayzliu) [reply examples/pytorch/geniepath/ppi.py:91]: ditto

File: examples/pytorch/geniepath/model.py [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+        self.depths = nn.ModuleList()
+        self.num_heads = num_heads
+        self.residual = residual
+        for i in range(num_layers):
+            self.breaths.append(GATConv(hid_dim, hid_dim, num_heads=num_heads, residual=residual))
+            self.depths.append(LSTM(hid_dim*2, hid_dim))
+
+    def forward(self, graph, x):
+        h = th.rand(1, x.shape[0], self.hid_dim).to(x.device)
+        c = th.rand(1, x.shape[0], self.hid_dim).to(x.device)
```
[the remark this answers is not in the document]
Comment (kayzliu) [reply examples/pytorch/geniepath/model.py:71]: Modified to zeros.

File: examples/pytorch/geniepath/model.py [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+
+class GeniePath(nn.Module):
+    def __init__(self, in_dim, out_dim, hid_dim=16, num_layers=2, num_heads=1, residual=False):
+        super(GeniePath, self).__init__()
+        self.in_dim = in_dim
+        self.out_dim = out_dim
+        self.hid_dim = hid_dim
+        self.num_layers = num_layers
+        self.linear1 = nn.Linear(in_dim, hid_dim)
+        self.linear2 = th.nn.Linear(hid_dim, out_dim)
```
[the remark this answers is not in the document]
Comment (kayzliu) [reply examples/pytorch/geniepath/model.py:33]: Deleted in the latest commit.

File: examples/pytorch/geniepath/model.py [hunk 6]
```diff
@@ -0,0 +1,83 @@
+import torch as th
+import torch.nn as nn
+from dgl.nn import GATConv
+from torch.nn import LSTM
+
+
+class GeniePathConv(nn.Module):
+    def __init__(self, in_dim, hid_dim, out_dim, num_heads=1, residual=False):
+        super(GeniePathConv, self).__init__()
+        self.in_dim = in_dim
```
[the remark this answers is not in the document]
Comment (kayzliu) [reply examples/pytorch/geniepath/model.py:10]: Deleted.

File: examples/pytorch/geniepath/model.py [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+        x = th.mean(x, dim=1)
+        x, (h, c) = self.depth_func(x.unsqueeze(0), (h, c))
+        x = x[0]
+        return x, (h, c)
+
+
+class GeniePath(nn.Module):
+    def __init__(self, in_dim, out_dim, hid_dim=16, num_layers=2, num_heads=1, residual=False):
+        super(GeniePath, self).__init__()
+        self.in_dim = in_dim
```
[the remark this answers is not in the document]
Comment (kayzliu) [reply examples/pytorch/geniepath/model.py:28]: ditto

File: examples/pytorch/geniepath/model.py [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+            x, (h, c) = layer(graph, x, h, c)
+        x = self.linear2(x)
+
+        return x
+
+
+class GeniePathLazy(nn.Module):
+    def __init__(self, in_dim, out_dim, hid_dim=16, num_layers=2, num_heads=1, residual=False):
+        super(GeniePathLazy, self).__init__()
+        self.in_dim = in_dim
```
[the remark this answers is not in the document]
Comment (kayzliu) [reply examples/pytorch/geniepath/model.py:55]: ditto

File: examples/pytorch/geniepath/ppi.py (hunk 3 above)
Comment (zhjwy9343) [reply examples/pytorch/geniepath/ppi.py:25]: ok

File: examples/pytorch/geniepath/ppi.py (hunk 1 above)
Comment (zhjwy9343) [reply examples/pytorch/geniepath/ppi.py:85]: ok

**Review by zhjwy9343 — APPROVED** [review]
LGTM

**Pull request merged**